File: dart/bin/lcd.dart
import 'package:args/args.dart';
import 'dart:io';
import 'dart:math';

class LCD {
  static const List<int> TABLE = const [119, 36, 93, 109, 46, 107, 123, 37, 127, 111];
  static const String H_SYM = "-";
  static const String V_SYM = "|";
  static const String S_SYM = " ";

  List<int> numbers;
  int segments;

  LCD (String input, int segments) {
    this.numbers = new List<int>();
    for (var c in input.runes) {
      numbers.add(int.parse(new String.fromCharCode(c)));
    }
    this.segments = segments;
  }

  bool test(int number, int position) {
    return (TABLE[number] & pow(2, position)) > 0;
  }

  void draw() {
    int offset = 0;
    for (int row = 0; row < 5; row++)
    {
      if (row % 2 == 0) {
        for (int number in numbers) {
           String symbol = test(number, row + offset) ? H_SYM : S_SYM;
           String symbols = "";
           for (int i = 0; i < segments; i++) {
             symbols += symbol;
           }
           stdout.write(S_SYM + symbols + S_SYM + S_SYM);
        }
        stdout.write("\n");
      } else {
        for (int i = 0; i < segments; i++) {
          for (int number in numbers) {
            for (int column = 0; column < 2; column++) {
              String symbol = test(number, row + column + offset) ? V_SYM : S_SYM;
              String spaces = "";
              for (int j = 0; j < (column == 0 ? segments : 1); j++) {
                spaces += S_SYM;
              }
              stdout.write(symbol + spaces);
            }
          }
          stdout.write("\n");
        }
        offset++;
      }
    }
  }
}

void main() {
  var parser = new ArgParser();
  parser.addOption('segments', abbr: 's', defaultsTo: '2');
  var results = parser.parse(new Options().arguments);
  if (results.rest.length > 0) {
    String input = results.rest.first;
    int segments = int.parse(results['segments']);
    LCD lcd = new LCD(input, segments);
    lcd.draw();
  } else {
    print("Usage: lcd [-s SEGMENTS] NUMBERS");
  }
}

File: lua/lcd.lua
LCD = {}
LCD.__index = LCD;

local TABLE = { 119, 36, 93, 109, 46, 107, 123, 37, 127, 111 }
local H_SYM = "-"
local V_SYM = "|"
local S_SYM = " "

function LCD.create(input, segments)
  local self = {}
  setmetatable(self, LCD);

  self.numbers = {}
  for i = 1, #input do
    self.numbers[i] = tonumber(input:sub(i,i))
  end
  self.segments = segments
  return self
end

function LCD:test(number, position)
  return bit32.band(TABLE[number + 1], 2 ^ position) > 0
end

function LCD:draw()
  local offset = 0
  for row = 0, 4 do
    if (row % 2 == 0) then
      for _, number in ipairs(self.numbers) do
        local symbol = self:test(number, row + offset) and H_SYM or S_SYM
        local symbols = ""
        for _ = 1, self.segments do
          symbols = symbols .. symbol
        end
        io.write(S_SYM .. symbols .. S_SYM .. S_SYM)
      end
      io.write('\n')
    else
      for _ = 1, self.segments do
        for _, number in ipairs(self.numbers) do
          for column = 0, 1 do
            local symbol = self:test(number, row + column + offset) and V_SYM or S_SYM
            local spaces = ""
            for _ = 1, (column == 0 and self.segments or 1) do
              spaces = spaces .. S_SYM
            end
            io.write(symbol .. spaces)
          end
        end
        io.write('\n')
      end
      offset = offset + 1
    end
  end
end

if #arg > 0 then
  local input
  local segments = 2
  if arg[1] == "-s" then
    segments = arg[2]
    input = arg[3]
  else
    input = arg[1]
  end
  local lcd = LCD.create(input, segments)
  lcd:draw()
else
  print("Usage: lua lcd.lua [-s SEGMENTS] NUMBER")
end
